src/lib.rs

```diff
@@ -0,0 +1,414 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A failpoints implementation for rust.
+
+extern crate rand;
+
+use std::str::FromStr;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Condvar, Mutex, RwLock, TryLockError};
+use std::thread;
+use std::time::{Duration, Instant};
+
+use rand::Closed01;
+
+/// Supported tasks.
+#[derive(Clone, Debug, PartialEq)]
+enum Task {
+    /// Do nothing.
+    Off,
+    /// Return the value.
+    Return(Option<String>),
+    /// Sleep for some milliseconds.
+    Sleep(u64),
+    /// Panic with the message.
+    Panic(Option<String>),
+    /// Print the message.
+    Print(Option<String>),
+    /// Sleep until other action is set.
+    Pause,
+    /// Yield the CPU.
+    Yield,
+    /// Busy waiting for some milliseconds.
+    Delay(u64),
+}
+
+#[derive(Debug)]
+struct Action {
+    task: Task,
+    freq: f32,
+    count: Option<AtomicUsize>,
+}
+
+impl PartialEq for Action {
+    fn eq(&self, hs: &Action) -> bool {
+        if self.task != hs.task || self.freq != hs.freq {
+            return false;
+        }
+        if let Some(ref lhs) = self.count {
+            if let Some(ref rhs) = hs.count {
+                return lhs.load(Ordering::Relaxed) == rhs.load(Ordering::Relaxed);
+            }
+        } else if hs.count.is_none() {
+            return true;
+        }
+        false
+    }
+}
+
+impl Action {
+    fn new(task: Task, freq: f32, max_cnt: Option<usize>) -> Action {
+        Action {
+            task: task,
+            freq: freq,
+            count: max_cnt.map(AtomicUsize::new),
+        }
+    }
+
+    fn get_task(&self) -> Option<Task> {
+        if let Some(ref cnt) = self.count {
+            let c = cnt.load(Ordering::Acquire);
+            if c == 0 {
+                return None;
+            }
+        }
+        if self.freq < 1f32 {
+            let Closed01(f) = rand::random::<Closed01<f32>>();
+            if f > self.freq {
+                return None;
+            }
+        }
+        if let Some(ref cnt) = self.count {
+            loop {
+                let c = cnt.load(Ordering::Acquire);
+                if c == 0 {
+                    return None;
+                }
+                if c == cnt.compare_and_swap(c, c - 1, Ordering::AcqRel) {
+                    break;
+                }
+            }
+        }
+        Some(self.task.clone())
+    }
+}
+
+fn partition(s: &str, pattern: char) -> (&str, Option<&str>) {
+    let mut splits = s.splitn(2, pattern);
+    (splits.next().unwrap(), splits.next())
+}
+
+impl FromStr for Action {
+    type Err = String;
+
+    /// Parse an action.
+    ///
+    /// `s` should be in the format `[p%][cnt*]task[(args)]`, `p%` is the frequency,
+    /// `cnt` is the max times the action can be triggered.
+    fn from_str(s: &str) -> Result<Action, String> {
+        let mut remain = s.trim();
```

Comment:
can we use a regexp to handle this? 

I guess may be easier here.

Reply:
I don't want to introduce a heavy and complicated dependency for this simple functionality.